src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
+    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ~ opt(dataType) ^^ {
+      case (r ~ isQuery ~ "(" ~ attrs ~ ")" ~ vType) => {
         val vars = attrs.zipWithIndex map { case(x, i) => Variable(x.name, r, i) }
         var types = attrs map { case(x) => x.t }
-        SchemaDeclaration(Attribute(r, vars, types), (isQuery != None))
+        val variableType = vType match {
+          case None => if (isQuery != None) Some(BooleanType) else None
+          case Some(s) => Some(s)
+        }
+        SchemaDeclaration(Attribute(r, vars, types), (isQuery != None), variableType)
       }
     }
 
+  def operator = "||" | "+" | "-" | "*" | "/" | "&" | "::"
+  val aggregationFunctions = Set("MAX", "SUM", "MIN", "ARRAY_ACCUM", "ARRAY_AGG")
+
+  // expressions
+  def expr : Parser[Expr] =
+    ( lexpr ~ operator ~ expr ^^ { case (lhs ~ op ~ rhs) => BinaryOpExpr(lhs, op, rhs) }
+    | lexpr
+    )
+
+  def lexpr : Parser[Expr] =
+    ( functionName ~ "(" ~ rep1sep(expr, ",") ~ ")" ^^ { 
+        case (name ~ _ ~ args ~ _) => FuncExpr(name, args, (aggregationFunctions contains name))
+      }
+    | constant ^^ { ConstExpr(_) }
+    | variableName ^^ { VarExpr(_) }
+    | "(" ~> expr <~ ")" 
+    )
 
   // TODO support aggregate function syntax somehow
-  def cqHead = relationName ~ "(" ~ repsep(variableName, ",") ~ ")" ^^ {
+  def cqHead = relationName ~ "(" ~ rep1sep(expr, ",") ~ ")" ^^ {
       case (r ~ "(" ~ variableUses ~ ")") =>
         Atom(r, variableUses.zipWithIndex map {
-          case(name,i) => Variable(name, r, i)
+          case (e,i) => ColumnExpr(e, r, i)
         })
     }
 
-  // TODO add conditional expressions for where clause
-  def cqConditionalExpr = failure("No conditional expression supported yet")
-  def cqBodyAtom: Parser[Atom] =
-    ( relationName ~ "(" ~ repsep(variableName, ",") ~ ")" ^^ {
-        case (r ~ "(" ~ variableBindings ~ ")") =>
-          Atom(r, variableBindings.zipWithIndex map {
-            case(name,i) => Variable(name, r, i)
-          })
+  // conditional expressions
+  def compareOperator = "LIKE" | ">" | "<" | ">=" | "<=" | "!=" | "=" | "IS" | "IS NOT"
+  def cond : Parser[Cond] = 
+    ( acond ~ (";") ~ cond ^^ { case (lhs ~ op ~ rhs) =>
+        BinaryOpCond(lhs, LogicOperator.OR, rhs)
+      }
+    | acond
+    )
+  def acond : Parser[Cond] = 
+    ( lcond ~ (",") ~ acond ^^ { case (lhs ~ op ~ rhs) => BinaryOpCond(lhs, LogicOperator.AND, rhs) }
+    | lcond
+    )
+  // ! has higher priority...
+  def lcond : Parser[Cond] = 
+    ( "!" ~> bcond ^^ { NegationCond(_) }
+    | bcond
+    )
+  def bcond : Parser[Cond] = 
+    ( expr ~ compareOperator ~ expr ^^ { case (lhs ~ op ~ rhs) =>
+        ComparisonCond(lhs, op, rhs)
       }
-    | cqConditionalExpr
+    | "[" ~> cond <~ "]"
     )
+
+  def cqBodyAtom: Parser[Atom] =
+    relationName ~ "(" ~ repsep(expr, ",") ~ ")" ^^ {
+      case (r ~ "(" ~ variableBindings ~ ")") =>
```

Comment:
Rename to `variableBindingsAndExprs` or `patterns`?

Eventually the grammar for the body columns should become *patterns* as well as one side (lhs) of equality conditions.  Patterns may be very similar to expr, but can be slightly different, e.g., having wildcards.